### DISCUSSION TITLE
Can't drag down the View Controller

### DISCUSSION THREAD

**Discussion by sasan-soroush** [LeoNatan/LNPopupController#466] [General] [answered]
Hi Leo, 
Hope you're doing awesome! 
When I present the View Controller from popup bar I can't drag it down, I have seen similar issues on this repo but I couldn't find out how to resolve this issue, 
I have a custom TabBarController & NavigationController, I don't know if it has to do anything with this issue, 
Although it is working fine when I add it to new projects just for testing so I guess it is sth wrong with my project,
I wonder what can cause this behavior, 

right now I just added a swipe gesture recognizer with `.down` direction to close the pop-up but I really like how it feels and looks on the original working version,

Also, The presented View Controller doesn't have any gesture recognizers when I try to print `view.gesturerecognizers`,
can it be related to this?

This is the core code for my custom tab bar controller:

```
class BaseTabBarController: UITabBarController {
    override func viewDidLoad() {
        super.viewDidLoad()

        let demoVC = FullScreenMusicPlayer()
        demoVC.popupItem.title = "Hello World"
        demoVC.popupItem.subtitle = "And a subtitle!"
        demoVC.popupItem.progress = 0.34
            
        self.presentPopupBar(withContentViewController: demoVC, animated: true, completion: nil)
    }
}
```

and this the View Controller that is being presented:

```
class FullScreenMusicPlayer: UIViewController {
    override func viewDidLoad() {
        super.viewDidLoad()
        
        view.backgroundColor = .red
    }
}
```

https://user-images.githubusercontent.com/31206210/123952180-a6553300-d973-11eb-9d17-a83b9fb66bfd.mov

Thanks in advance

**Comment by LeoNatan** [comment]
Hello,
Have you overridden the `viewForPopupInteractionGestureRecognizer` property? That’s the only reason I can think of that would make it not work, since the controller in the video seems a simple one. Or disabled interaction with `LNPopupInteractionStyleNone`.

**Reply by sasan-soroush** [reply]
Would you please tell me how exactly should I do that? 
Where should I override `viewForPopupInteractionGestureRecognizer` and what should I override it with?

I tried overriding it on **Presented View Controller** and returned `self.view` but it didn't change anything, 
And I know nothing about `LNPopupInteractionStyleNone`.

**Reply by LeoNatan** [reply]
You should not override it unless you want a specific view to handle interaction.

**Reply by sasan-soroush** [reply]
I want the presented view controller's view to handle this interaction as it does by default.

What can I do with the `LNPopupInteractionStyleNone` property?

When installing this package through SPM it asks me to choose between the dynamic and static version of the lib, which one should I choose? (I actually tried both but I am not sure what is the correct choice)

**Reply by LeoNatan** [reply]
Yes, the default behavior is that the interaction gesture recognizer is added to the controller's view. `LNPopupInteractionStyleNone` is to **disable** interaction, which is not what you want. Dynamic/static shouldn't make a difference. Static embeds the framework inside your binary, whereas dynamic is compiled as a dylib which is loaded on launch.

Perhaps update your question with the code for your content controller.

**Reply by sasan-soroush** [reply]
Sure, Updated the question with the codes.

**Reply by LeoNatan** [reply]
There is no reason that code shouldn’t work. Check out the example project, where similar flows happen.

**Reply by sasan-soroush** [reply]
I noticed something weird,

When I print the gesture recognizers on your example project (Demo - Music Scene (Swift)) with the following code,

```
tabBarController?.popupBar.gestureRecognizers?.forEach({ rec in
       print("GESTURE RECOGNIZER: ", rec)
})
```

I get this:

```
GESTURE RECOGNIZER:  <LNPopupOpenTapGestureRecognizer: 0x7fcaad54f760; baseClass = UITapGestureRecognizer; state = Possible; view = <LNPopupBar 0x7fcaad61e770>; target= <(action=_popupBarTapGestureRecognized:, target=<LNPopupController 0x7fcaad710560>)>>
GESTURE RECOGNIZER:  <LNPopupLongPressGestureRecognizer: 0x7fcaad54fa80; baseClass = UILongPressGestureRecognizer; state = Possible; cancelsTouchesInView = NO; delaysTouchesEnded = NO; view = <LNPopupBar 0x7fcaad61e770>; target= <(action=_popupBarLongPressGestureRecognized:, target=<LNPopupController 0x7fcaad710560>)>; numberOfTapsRequired = 0>
GESTURE RECOGNIZER:  <LNPopupInteractionPanGestureRecognizer: 0x7fcaad764860; baseClass = UIPanGestureRecognizer; state = Possible; view = <LNPopupBar 0x7fcaad61e770>; target= <(action=_popupBarPresentationByUserPanGestureHandler:, target=<LNPopupController 0x7fcaad710560>)>>
```

but when I print it on my project 

```
demoVC.popupBar.gestureRecognizers?.forEach({ rec in
      print("GESTURE RECOGNIZER: ", rec)
})
```

I get this: 

```
GESTURE RECOGNIZER:  <LNPopupOpenTapGestureRecognizer: 0x7ff45b84d530; baseClass = UITapGestureRecognizer; state = Possible; view = <LNPopupBar 0x7ff459f50240>; target= <(action=_popupBarTapGestureRecognized:, target=<LNPopupController 0x7ff45b84e2d0>)>>
GESTURE RECOGNIZER:  <LNPopupLongPressGestureRecognizer: 0x7ff45b858060; baseClass = UILongPressGestureRecognizer; state = Possible; cancelsTouchesInView = NO; delaysTouchesEnded = NO; view = <LNPopupBar 0x7ff459f50240>; target= <(action=_popupBarLongPressGestureRecognized:, target=<LNPopupController 0x7ff45b84e2d0>)>; numberOfTapsRequired = 0>
```

as you see in my project there is no UIPanGestureRecognizer, What are the possible reasons for that?

**Reply by LeoNatan** [reply]
The gesture recognizer is added to the popup bar here:
https://github.com/LeoNatan/LNPopupController/blob/e76d558de57279572069ed79841b6cf0c91d67f2/LNPopupController/LNPopupController/Private/LNPopupController.m#L316

and remove here:
https://github.com/LeoNatan/LNPopupController/blob/e76d558de57279572069ed79841b6cf0c91d67f2/LNPopupController/LNPopupController/Private/LNPopupController.m#L332

I see no reason why this code wouldn't work.
If you wish, you can put breakpoints there and see why it's not working.

**Reply by sasan-soroush** [reply]
I put breakpoints on these lines as you told, 
when I open the popup bar it removes the gesture recognizer (line 332)
and when I close the presented view controller it will add it again (line 316)

I attached the screen recording,

Thanks.


https://user-images.githubusercontent.com/31206210/124353109-79528b80-dbd2-11eb-9d21-f42d3d3b6937.mov

**Reply by LeoNatan** [reply]
That is correct. It adds the gesture recognizer to the bar when the popup is closed and removes it when open. It should add it to your controller’s view when open.

**Reply by sasan-soroush** [reply]
I see,
so the actual problem is that it doesn't add the gesture recognizer to my presented view controller's view, 
How can I fix that? 

Although when I tried printing gesture recognizers in the presented view controller's view inside your sample project (Demo - Music Scene (Swift)) it doesn't print anything, so I guess I am looking for it in the wrong place? I don't know.

**Reply by LeoNatan** [reply]
If you print the gesture recognizers of the player view **when the popup is open**, you should see an instance of `LNPopupInteractionPanGestureRecognizer`, which is the class of the interaction gesture recognizer.

**Reply by sasan-soroush** [reply]
Alright, but how should I fix it for my project? I didn't do anything to prevent it from adding the gesture recognizer to the view.
is there any way to add it myself or at least replicate it?

**Reply by iDevelopper** [reply]
@sasan-soroush ,

> Although when I tried printing gesture recognizers in the presented view controller's view inside your sample project (Demo - Music Scene (Swift)) it doesn't print anything, so I guess I am looking for it in the wrong place? I don't know.

Where do you try to print the gesture recognizers. Even if you try to print them in viewDidAppear, it is too early. The gesture recognizer is added after.

**Reply by sasan-soroush** [reply]
Oh thanks, I didn't know that, when/where should I print it to get the correct result?

**Reply by iDevelopper** [reply]
Add a view controller to your storyboard of FullScreenMusicPlayer class (don't forget the storyboard identifier), add a UIButton to the view to print the gesture recognizers, and connect the button to an IBAction:

```swift
    @IBAction func printGestureRecognizer(_ sender: UIButton) {
        print("Gestures: \(String(describing: self.view.gestureRecognizers))")
    }
```
In BaseTabBarController.swift:
```swift
    override func viewDidLoad() {
        super.viewDidLoad()
        
        if let demoVC = self.storyboard?.instantiateViewController(withIdentifier: "FullScreenMusicPlayer") as? FullScreenMusicPlayer {
            demoVC.popupItem.title = "Hello World"
            demoVC.popupItem.subtitle = "And a subtitle!"
            demoVC.popupItem.progress = 0.34
            
            self.presentPopupBar(withContentViewController: demoVC, animated: true, completion: nil)
        }
    }
```

**Reply by LeoNatan** [reply]
Right, that’s a good idea. Then tap the button **when you open the popup**.

**Reply by iDevelopper** [reply]
You can also print the container and the interaction style:
```swift
    @IBAction func printGestureRecognizer(_ sender: UIButton) {
        print("Container: \(String(describing: self.popupPresentationContainer))")
        print("Interaction style: \(String(describing: self.popupPresentationContainer?.popupInteractionStyle.rawValue))")
        print("Gestures: \(String(describing: self.view.gestureRecognizers))")
    }
```

**Reply by iDevelopper** [reply]
And if you want the progress bar to be displayed you have to set the style (default is none):
```swift
            self.popupBar.progressViewStyle = .top
            demoVC.popupItem.progress = 0.34
```

**Reply by sasan-soroush** [reply]
Thanks for the solution! I think we are one step closer now,
I did what you said and figured that it does add the gesture recognizer to the presented view controller's view
Here is what it prints: 
```
Container: Optional(<MyProject.BaseTabBarController: 0x7f847c08c600>)
Interaction style: Optional(0)
Gestures: Optional([<LNPopupInteractionPanGestureRecognizer: 0x7f847bcf2050; baseClass = UIPanGestureRecognizer; state = Possible; view = <UIView 0x7f847bdf7080>; target= <(action=_popupBarPresentationByUserPanGestureHandler:, target=<LNPopupController 0x7f847cf240c0>)>>])
```
So I guess the only problem here is that the pan gesture recognizer doesn't work
I added the following codes but it's not working.

```
func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer, shouldRecognizeSimultaneouslyWith otherGestureRecognizer: UIGestureRecognizer) -> Bool {
    return true
}

override var viewForPopupInteractionGestureRecognizer: UIView {
    return view
}
```

**Reply by iDevelopper** [reply]
Is the UIView is the FullScreenMusicPlayer' view? What is the delegate of the LNPopupInteractionPanGestureRecognizer?
```swift
        print("Delegate: \(String(describing: self.popupPresentationContainer!.popupContentView.popupInteractionGestureRecognizer.delegate))")
```

**Reply by sasan-soroush** [reply]
it prints `nil`, but it also prints `nil` in another project that pan gesture works.

**Reply by iDevelopper** [reply]
Is the method _popupBarPresentationByUserPanGestureHandler_began: called when you try to dismiss the popup by dragging down? If so, can you put some break points at each return; statement? Also in _popupBarPresentationByUserPanGestureHandler_changed:

**Reply by sasan-soroush** [reply]
`_popupBarPresentationByUserPanGestureHandler_began`  is getting called but it doesn't hit any of those returns,
`_popupBarPresentationByUserPanGestureHandler_changed` however, never gets called, it's the only difference it has with the example project.


https://user-images.githubusercontent.com/31206210/124706387-26741f00-dec5-11eb-888b-fdf9e56caeb7.mov

it hits the `_popupBarPresentationByUserPanGestureHandler_changed` without going into changed state

**Reply by LeoNatan** [reply]
If you can create a **small** project where this reproduces, I can take a look why the gesture recognizer is failing. My instinct says user error, rather than framework failure, but I can at least give you an idea why.

**Reply by iDevelopper** [reply]
> right now I just added a swipe gesture recognizer with .down direction to close the pop-up but I really like how it feels and looks on the original working version

I can reproduce your issue with a swipe gesture added. Have you removed that gesture from view?

**Reply by sasan-soroush** [reply]
Yeah, I just comment that out when I am trying to figure out what is going wrong, 
at first, I assumed that maybe another library is adding another gesture recognizer but when I print all of them as you said, there is no other gesture recognizer.

**Reply by LeoNatan** [reply, marked as answer]
Wait, just remembered from another discussion. Do you use Mixpanel? They add a gesture recognizer to the window (🤦‍♂️), which breaks the framework gesture recognition logic.

**Reply by sasan-soroush** [reply]
Wow finally we know why it was behaving like this, I was going crazy,
You're absolutely right, Mixpanel was the reason for this issue, do you know how should I deal with that ?

**Reply by LeoNatan** [reply]
Remove Mixpanel? I have no idea what it does, or why it has to add a gesture recognizer to the window. If you must use it, please open an issue with them. Thanks

**Reply by sasan-soroush** [reply]
Sure!
I feel relieved because of knowing what is wrong now!
I will talk to other developers to see if we can remove this or disable the gesture recognizer, 
BTW, thank you so much for your patience and help ! that means a lot.
